src/main/java/com/aws/greengrass/util/platforms/unix/UnixPlatform.java

```diff
@@ -376,75 +370,89 @@ public void addUserToGroup(String user, String group) throws IOException {
-                    logger.atTrace().setEventType(SET_PERMISSIONS_EVENT).kv(PATH, path)
-                            .kv("owner", permission.getOwnerUser()).log();
-                    view.setOwner(userPrincipal);
-                    if (groupPrincipal != null) {
-                        logger.atTrace().setEventType(SET_PERMISSIONS_EVENT).kv(PATH, path)
-                                .kv("group", permission.getOwnerGroup()).log();
-                        view.setGroup(groupPrincipal);
-                    }
-                    return null;
-                };
-            }
+        if (groupPrincipal != null) {
+            logger.atTrace().setEventType(SET_PERMISSIONS_EVENT).kv(PATH, path).kv("group", groupPrincipal.toString())
+                    .log();
+            view.setGroup(groupPrincipal);
         }
+    }
 
+    @Getter
+    public static class PosixFileSystemPermissionView extends FileSystemPermissionView {
+        private Set<PosixFilePermission> posixFilePermissions;
 
-        // noop function that does not change the file mode
-        CrashableFunction<PosixFileAttributeView, Void, IOException> setMode = (p) -> null;
+        public PosixFileSystemPermissionView(FileSystemPermission permission) {
+            super();
+            posixFilePermissions = posixFilePermissions(permission);
+        }
 
-        if (options.contains(Option.SetMode)) {
-            Set<PosixFilePermission> perms = permission.toPosixFilePermissions();
-            setMode = (view) -> {
-                logger.atTrace().setEventType(SET_PERMISSIONS_EVENT).kv(PATH, path).kv("perm",
-                        PosixFilePermissions.toString(perms)).log();
-                view.setPermissions(perms);
-                return null;
-            };
-        }
-        final CrashableFunction<PosixFileAttributeView, Void, IOException> setModeFunc = setMode;
-        final CrashableFunction<PosixFileAttributeView, Void, IOException> setOwnerFunc = setOwner;
-        if (options.contains(Option.Recurse)) {
-            Files.walkFileTree(path, new SimpleFileVisitor<Path>() {
-                @Override
-                public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
-                    PosixFileAttributeView view = Files.getFileAttributeView(dir, PosixFileAttributeView.class,
-                            LinkOption.NOFOLLOW_LINKS);
-                    setModeFunc.apply(view);
-                    setOwnerFunc.apply(view);
-                    return FileVisitResult.CONTINUE;
-                }
+        /**
+         * Convert to a set of PosixFilePermissions for use with Files.setPosixFilePermissions.
+         *
+         * @param permission Permission to convert
+         * @return Set of permissions
+         */
+        public static Set<PosixFilePermission> posixFilePermissions(FileSystemPermission permission) {
+            Set<PosixFilePermission> ret = new HashSet<>();
```

Comment:
I think Rob mean to turn this into an EnumSet instead of HashSet, can you do that instead of the copy stuff at the end?

Reply:
My bad. I understood what Rob meant but misunderstood how to use EnumSet. I will fix it in the next commit.